pkg/resource/role/sdk.go

```diff
@@ -595,3 +532,36 @@ func (rm *resourceManager) terminalAWSError(err error) bool {
 		return false
 	}
 }
+
+// getImmutableFieldChanges returns list of immutable fields from the
+func (rm *resourceManager) getImmutableFieldChanges(
+	delta *ackcompare.Delta,
+) []string {
+	var fields []string
+	if delta.DifferentAt("Spec.Name") {
+		fields = append(fields, "Name")
+	}
+
+	return fields
+}
+
+// newUpdateRequestPayload returns an SDK-specific struct for the HTTP request
+// payload of the Update API call for the resource
+func (rm *resourceManager) newUpdateRequestPayload(
+	ctx context.Context,
+	r *resource,
+) *svcsdk.UpdateRoleInput {
+	res := &svcsdk.UpdateRoleInput{}
+
+	if r.ko.Spec.Description != nil {
+		res.SetDescription(*r.ko.Spec.Description)
+	}
+	if r.ko.Spec.MaxSessionDuration != nil {
+		res.SetMaxSessionDuration(*r.ko.Spec.MaxSessionDuration)
+	}
+	if r.ko.Spec.Name != nil {
+		res.SetRoleName(*r.ko.Spec.Name)
+	}
```

Comment:
Because you've used a custom update method, the immutability guarantee code is no longer generated for `sdkUpdate` which means this block of code here actually violates the immutability guarantee.

This is generally why I I don't care for the custom update method complete override. I think it would be better to just add this to a `sdk_update_pre_build_request.go.tpl` file:

```go
	if delta.DifferentAt("Spec.Policies") {
		err = rm.syncPolicies(ctx, desired, latest)
		if err != nil {
			return nil, err
		}
	}
	if delta.DifferentAt("Spec.Tags") {
		err = rm.syncTags(ctx, desired, latest)
		if err != nil {
			return nil, err
		}
	}
	if delta.DifferentAt("Spec.PermissionsBoundary") {
		err = rm.updateRolePermissionsBoundary(ctx, desired)
		if err != nil {
			return nil, err
		}
	}
	if !delta.DifferentExcept("Spec.Tags", "Spec.Policies", "Spec.PermissionsBoundary") {
		return desired, nil
	}
```

and get rid of the custom update method. If you regenerate the controller again after doing that, I think you'll find that the immutability guarantee code will be generated for you properly.

Reply:
Can't we just add the immutability handler manually to the custom update code? The problem with injecting code to pre/post build request is that we'll never prevent the controller from making `Update{Resource}` calls: [This block](https://github.com/aws-controllers-k8s/iam-controller/blob/main/pkg/resource/group/sdk.go#L252-L258) will always be executed even if we only modify the policies

Reply:
> Can't we just add the immutability handler manually to the custom update code? The problem with injecting code to pre/post build request is that we'll never prevent the controller from making `Update{Resource}` calls: [This block](https://github.com/aws-controllers-k8s/iam-controller/blob/main/pkg/resource/group/sdk.go#L252-L258) will always be executed even if we only modify the policies

No, it won't... see my code above. This code:

```go
	if !delta.DifferentExcept("Spec.Tags", "Spec.Policies", "Spec.PermissionsBoundary") {
		return desired, nil
	}
```

will prevent that.

Reply:
Ohhh... fair enough that's a good one. And indeed way better solution.

Reply:
Fixed